File: ztm.js
/**
 * Skrypt generujący dane o przystankach końcowych oraz trasach linii tramwajowych i autobusowych
 */
var fs = require('fs'),
	bot = require('nodemw'),
	client = new bot('config.js');

function sortObject(obj) {
	var sorted = {},
		keys = Object.keys(obj);

	keys.sort();

	for (var i=0, len = keys.length; i<len; i++) {
		sorted[keys[i]] = obj[keys[i]];
	}

	return sorted;
}

function parseTimetable(page, line) {
	var matches = page.match(routeRegExp);

	if (matches) {
		// parsuj trasę + usuń przystanki końcowe
		var streets = matches[1].split(/-|–/).
			slice(1, -1).
			map(function(item) {
				return item.trim();
			});

		console.log('#' + line + ': ' + streets.join(', '));

		streets.forEach(function(street) {
			// ignoruj ronda
			if (street.indexOf('Rondo') > -1) {
				return;
			}

			// rozwiń skróty + małe poprawki
			street = street.
			replace("Al.", 'Aleja').
				replace(/[śŚ]w\./, 'Święty').
				replace('ŚwiętyMarcin', 'Święty Marcin').
				replace('Świety', 'Święty').
				replace('Piasnicka', 'Piaśnicka').
				replace('Os. ', 'Osiedle ');
		
			if (street.indexOf('28 Czerwca') === 0) {
				street = '28 Czerwca 1956 r.';
			}

			if (street === '') {
				return;
			}

			ulice[street] = ulice[street] || [];

			if (ulice[street].indexOf(line) === -1) {
				ulice[street].push(line);

				// sortuj nr linii jako wartości liczbowe
				ulice[street].sort(function(a,b) {
					return (typeof a == "number") ? a - b : 1;
				});
			}
		});

		// aktualizuj "bazę"
		fs.writeFileSync('db/ztm-ulice.json', JSON.stringify(sortObject(ulice)));
	}

	// przystanki na trasie
	var stops = page.match(czasPrzejazduRegExp);

	if (stops) {
		// liczba przystanków
		linie[line].przystanki = stops.length + 1 /* przystanek początkowy */;

		// czas przejazdu
		var czas = stops.pop().substr(4);
		czas = parseInt(czas, 10);

		if (czas > 0) {
			linie[line].czas = czas;
			console.log('#' + line + ': ' + czas + ' min / ' + (stops.length + 1) + ' przystanków');
		}
	}

	// aktualizuj "bazę"
	fs.writeFileSync('db/ztm-linie.json', JSON.stringify(linie));
}

var linie = {},
 	ulice = {},
	timetableRegExp = /<a href='(timetable.html[^']+)'>/,
	timetableLastRegExp = /\<a href='(timetable.html.php[^']+)'>[^>]+<\/a><\/li><\/ul><\/div>/,
	routeRegExp = /<div id='descriptions'><p>([^<]+)/,
	petleRegExp = />([^<]+)<\/a><\/li><\/ul>/g,
	czasPrzejazduRegExp = /><b>\d+<\/b>/g,
	strefyRegExp = /<li class=\Szone(\S)/g;

var l, lines = [];

client.fetchUrl('http://www.ztm.poznan.pl/gtfs-ztm/routes_by_name.html.php?dbname=gtfs').then(function(page) {
	var links = page.match(/<a class='[^']+' href='[^']+'/g) || [],
		lines = [];

	links.forEach(function(link) {
		lines.push({
			no: link.match(/route_name=([^&]+)/)[1],
			agency: link.match(/agency_name=([^']+)/)[1]
		});
	});

	//console.log(links); console.log(lines); process.exit(1);

lines.forEach(function(lineData) {
	var line = lineData.no;

	line = parseInt(line, 10) || line;

	// autobusy za tramwaj / linia turystyczna
	if (line[0] == 'T' || line === '0') {
		return;
	}

	var url ='http://www.ztm.poznan.pl/gtfs-ztm/route_directions.html.php?route_name=' + line + '&agency_name=' + lineData.agency;

	// nowe wersje rozkładu (przed zmianami tras)
	url += '&dbname=gtfs';

	console.log("Linia " + line + " (" + lineData.agency + ")");

	client.fetchUrl(url).then(function(page) {
		// pobierz rozkład jazdy -> trasa w obie strony
		var timetableUrl = page.match(timetableRegExp),
			timetableLastUrl = page.match(timetableLastRegExp);
		
		// przygotuj dane linii
		linie[line] = {
			petle: [],
			strefy: [],
			przystanki: false,
			czas: false,
			agency: lineData.agency.replace(/_/g, ' ')
		};

		if (timetableUrl) {
			client.fetchUrl('http://www.ztm.poznan.pl/gtfs-ztm/' + timetableUrl[1] + '&dbname=gtfs').then(function(page) {
				parseTimetable(page, line);
			});
		}

		if (timetableLastUrl) {
			client.fetchUrl('http://www.ztm.poznan.pl/gtfs-ztm/' + timetableLastUrl[1] + '&dbname=gtfs').then(function(page) {
				parseTimetable(page, line);
			});
		}

		// parsuj pętle
		var matches = page.match(petleRegExp) || []

		matches.forEach(function(match) {
			var stop = match.substr(1, match.length - 15).trim();

			// strefy taryfowe
			stop = stop.replace('(B)', '');

			// ucfirst()
			stop = stop[0] + stop.substring(1).toLowerCase();

			// porządki
			stop = stop.
				trim().
				replace("Os. sobieskiego", "Osiedle Jana III Sobieskiego").
				replace(/pl\. /i, "Plac ").
				replace(/os\. /i, "Osiedle ").
				replace('Os.wichrowe', 'Osiedle Wichrowe').
				replace("os.", "Osiedle ").
				// os. batorego ii -> os. batorego II
				replace(/\si+$/g, function(match) {
					return match.toUpperCase();
				});

			// ucfirst()
			stop = stop.replace(/[\s|\/][\w\W]/g, function(match) {
				return match.toUpperCase()
			});

			// końcówki
			linie[line].petle.push(stop);
		});

		// parsuj strefy taryfowe
		matches = page.match(strefyRegExp) || [];

		matches.forEach(function(match) {
			var strefa = match.substr(-1).toUpperCase();

			if (linie[line].strefy.indexOf(strefa) === -1) {
				linie[line].strefy.push(strefa);
				linie[line].strefy = linie[line].strefy.sort();
			}
		});

		console.log('#' + line + ': ' + JSON.stringify(linie[line]));
		
		// aktualizuj "bazę"
		fs.writeFileSync('db/ztm-linie.json', JSON.stringify(linie));
	});
});
});

File: ztm.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Skrypt generujący dane dla skryptów aktualizujących
dane o liniach tramwajowych i autobusowych

db/ztm-linie.json:

  "1": {
    "petle": [
      "Junikowo",
      "Franowo"
    ],
    "strefy": [
      "A"
    ],
    "przystanki": 31,
    "czas": 41,
    "agency": "ZTM MPK"
  }

db/ztm-ulice.json:

    "Żegrze": [
        1,
        4,
        5,
        13,
        17,
        55,
        66,
        201,
        245
    ]

db/ztm-stops.lua:

local database = {
    ["LUGA01"] = { "54" },
    ["LUGA02"] = { "54", "62", "244" },
    ["PILS01"] = { "52", "55", "62", "65" },
    ["OSWI01"] = { "52", "55", "62", "65" },
    ["SZII02"] = { "52", "74", "84" },
}

return database

"""

import csv
import json

# czytaj ze źródeł
with open("sources/ztm-routes.json") as f:
    routes = json.load(f)

with open("sources/ztm-operators.json") as f:
    try:
        operators = json.load(f)
    except ValueError:
        operators = {'lines': []}

# generuj dane o liniach i przystankach
lines = {}
stops = {}

for line in routes['lines']:
    try:
        line['name'] = int(line['name'])
    except:
        pass

    # rejestruj linię
    lines[line['name']] = {
        "typ": line['typ'],
        "petle": line['petle'],
        "przystanki": line['przystanki'],
    }

    # rejestruj linię zatrzymujące się na poszczególnych przystankach
    for stop in line['przystankiSymbole'].split(','):
        if stop not in stops:
            stops[stop] = []

	line_name = str(line['name'])

	if line_name not in stops[stop]:
	        stops[stop].append(line_name)

# generuj dane o przystankach (format dla skryptów LUA)
with open("db/ztm-stops.lua", "w") as lua:
    lua_lines = ['    ["%s"] = { "%s" }' % (stop, '", "'.join(stops[stop])) for stop in sorted(stops.keys())]

    lua.write('local database = {\n%s}\n\nreturn database\n' % ',\n'.join(lua_lines))

# typ / operator linii + rozkład jazdy
for line in operators['lines']:
    try:
        line['name'] = int(line['name'])
    except:
        pass

    if line['name'] in lines:
        if line['night'] is True:
            lines[line['name']]['night'] = True

        # operator
        lines[line['name']]['agency'] = line['operator'].replace('ZTM_', '')

        # link do rozkładu jazdy
        lines[line['name']]['rozklad'] = \
            'http://ztm.poznan.pl/komunikacja/rozklad/#/kierunki/%s/linia/%s' % (line['operator'], line['name'])

# TODO: inni operatorzy
# * Kombus - http://www.kombus.pl/komunikacja

# generuj plik JSON
with open("db/ztm-linie.json", "w") as out:
    json.dump(lines, out, indent=2, separators=(',', ':'), sort_keys=True)

# generuj dane o ulicach
stops_to_street = {}
streets = {}

with open("sources/ztm-stops.tsv", "r") as f:
    stops = csv.reader(f, delimiter="\t")

    for stop in stops:
        # tylko przystanki w Poznaniu
        if stop[4] != 'Poznań':
            continue

        # ID przystanku -> ulica / lokalizacja
        stops_to_street[stop[0]] = stop[5]

for line in routes['lines']:
    entry = []
    stops = line['przystankiSymbole'].split(',')

    for stop in stops:
        if stop in stops_to_street:
            street = stops_to_street[stop]

            if street not in streets:
                streets[street] = set()

            streets[street].add(line['name'])

for street in streets:
    streets[street] = sorted(list(streets[street]))

with open("db/ztm-ulice.json", "w") as out:
    json.dump(streets, out, indent=2, separators=(',', ': '), sort_keys=True)
